designs/0035-cmdstanrs.md

```diff
@@ -0,0 +1,348 @@
+out in Scheme, why I am not sure.
+
+### Ergonomics
+
+Philosophy:
+- pay (in LOC) for only what you need.
+- minimize differences between naming of the types and fields (see
+  below) in the Rust implementation and CmdStan.
+
+The builder pattern will be implemented for each `struct`, and for
+each `enum` variant (excluding unit variants). This enables the user
+to supply only the arguments for which they desire non-default
+values. This leads to succinct code when one needs only the defaults
+([example](https://github.com/andrewjradcliffe/cmdstan-rs/blob/main/examples/bernoulli-many/main.rs)).
+
+#### A side effect of strong typing
+
+With [company-mode](https://github.com/company-mode/company-mode) and
+[eglot-mode](https://github.com/joaotavora/eglot)
+([lsp-mode](https://github.com/emacs-lsp/lsp-mode/) also works) in
+Emacs 28.2, one can view options at each node in the argument tree by
+code that looks something like the following:
+
+```rust
+ArgumentTree::builder(). // hover on the `.`
+```
+
+If one has a Rust language server and completion support in their
+editor, this is a free side effect. Whether it will help anyone
+is uncertain.
+
+### Coverage
+
+The objective is for the interface to cover all options which can be
+passed to a compiled Stan program, that is, all methods and all
+options for said methods.
+
+## Separation of concerns
+
+Other than the argument tree support, the interface proposed is very
+simple: the user can compile a Stan program, call it with arguments,
+get basic information from the output, and call
+`diagnose`/`stansummary`.
+
+Below, I provide the rationale for exclusion of two aspects. My
+judgment is that they are useful, but are best developed separately.
+
+### Serialization
+
+It is trivial to provide a function such as 
+```rust
+fn write_json<T: Serialize>(data: &T, file: &Path) {}
+```
+but it serves no purpose -- it does not enforce the conventions
+adopted for representing Stan data types in JSON (e.g. matrices
+represented as a vector of *row* vectors, not a vector of *column*
+vectors), hence, would likely lead to unexpected (and potentially
+silent!) errors.
+
+In order to develop a serializer which respects the conventions for
+Stan data types, one would need to declare conventions for the mapping
+of Rust data types to Stan data
+types. [serde_json](https://github.com/serde-rs/json) would be nice to
+use, but has some incompatibilities (Rust tuple is represented as an
+array, rather than an object).
+
+Moreover, to represent matrices, complex numbers, etc., one would need
+to support types from the crate ecosystem since the standard library
+lacks these -- [nalgebra](https://github.com/dimforge/nalgebra) and
+[num-complex](https://github.com/rust-num/num-complex) are reasonable
+choices, but nonetheless represent decisions to be made!
+
+From a design perspective, this is a great place to defer to the user,
```

Comment:
For ease of implementation this is true, but the list of above subtleties (to me) make a loud argument in the opposite direction: this is something _hard_, and the obvious approach will easily lead to incorrect results, so it should be handled by the library (eventually?)

Reply:
Alas, yes, it should be handled by the library. It could be gated by a feature, so as to delay the need for an implementation for the initial version. As additional dependencies would likely be introduced, a feature gate is probably appropriate anyway.